src/utils/govPayMetadata.test.ts

```diff
@@ -0,0 +1,119 @@
+  },
+  { key: "key_number", value: "@number" },
+  { key: "key_object", value: "@object" },
+  { key: "key_boolean", value: "@boolean" },
+  {
+    key: "key_someValueNotSetInPassport",
+    value: "@someValueNotSetInPassportolean",
+  },
+];
+
+const mockPassport: Passport = {
+  data: {
+    string: "agent",
+    stringGreaterThan100Chars:
+      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus volutpat ex a tempor facilisis. Nullam eget mi velit.",
+    stringArray: ["new.building", "new.solarPanel"],
+    stringArrayGreaterThan100Chars: [
+      "Lorem ipsum dolor sit amet",
+      "consectetur adipiscing eli",
+      "Phasellus volutpat ex a tempor facilisis",
+      "Nullam eget mi velit",
+    ],
+    number: 123,
+    object: { someGeoJSON: { abc: 123 } },
+    boolean: true,
+  },
+};
+
+describe("transforming the payload to the correct format for GovPay", () => {
+  it("handles variables set in the Editor", () => {
+    const result = formatGovPayMetadata(mockMetadata, mockPassport);
+    expect(result).toMatchObject({
+      firstKey: "firstValue",
+      secondKey: "secondValue",
+    });
+  });
+
+  it("handles variables set in the Passport", () => {
+    const result = formatGovPayMetadata(mockMetadata, mockPassport);
+    expect(result).toMatchObject({
+      key_string: "agent",
+      key_boolean: true,
+    });
+  });
+});
+
+describe("validating data", () => {
+  const formatted = formatGovPayMetadata(mockMetadata, mockPassport);
+
+  const get = (testKey: string) => [
+    mockPassport.data[testKey], // Expected
+    formatted[`key_${testKey}`], // Result
+  ];
+
+  it("accepts strings", () => {
+    const [expected, result] = get("string");
+    // Value parsed as-is, no changes made
+    expect(result).toBe(expected);
+  });
+
+  it("accepts numbers", () => {
+    const [expected, result] = get("number");
+    // Value parsed as-is, no changes made
+    expect(result).toBe(expected);
+  });
+
+  it("accepts booleans", () => {
+    const [expected, result] = get("boolean");
+    // Value parsed as-is, no changes made
+    expect(result).toBe(expected);
+  });
+
+  it("converts string arrays to string", () => {
+    const [_, result] = get("stringArray");
+    const expected = "new.building, new.solarPanel";
+    // Array of strings joined to single string
+    expect(result).toBe(expected);
+  });
+
+  it("returns an error for undefined values", () => {
+    const [_, result] = get("someValueNotSetInPassport");
+    // Error passed to GovPay API as string, payment allows to proceed
+    expect(result).toMatch(
+      /Error: Invalid metadata value for "key_someValueNotSetInPassport" set in PlanX/,
+    );
+  });
+
+  it("returns an error for objects", () => {
+    const [_, result] = get("object");
+    // Error passed to GovPay API as string, payment allows to proceed
+    expect(result).toMatch(
+      /Error: Invalid metadata value for "key_object" set in PlanX/,
+    );
+  });
```

Comment:
I'll be curious how quickly this one comes up as a limitation - specifically anticipating `_address.uprn` - but very happy for it to throw an error initially !

Reply:
Yeah this is fair, it may well be something users ask for / expect. It's really just hard coded variables like `_address` that are objects really I think.

This could actually be allowed with recursion here - I'll take a look, it might actually not be a big change to allow this to work.

Reply:
Yeah also curious if / when this comes up!

The main issue we face here is the issue of key paths when accessing passport variables within objects, when the passport is a combination a of nested and flattened structure.

For example, if we wanted the value for `_address.uprn` we can't say `passport.data["_address.uprn"]`, we'd need to do `passport.data["_address"]["uprn"]`. Likewise, if we use a key path (like we do with the getters in the `Passport` class) we'd need to know where to split `_address.uprn`, but not `proposal.projectTypes`.

It's certainly a solvable problem, (check for existence of value, iterate over key splits) but not a trivial one. We'll need to tackle this when looking at embedding passport variables more widely within the Editor.